Sparkle/SUBinaryDeltaApply.m

```diff
@@ -15,38 +15,84 @@
 #include <stdlib.h>
 #include <xar/xar.h>
 
-static void applyBinaryDeltaToFile(xar_t x, xar_file_t file, NSString *sourceFilePath, NSString *destinationFilePath)
+static BOOL applyBinaryDeltaToFile(xar_t x, xar_file_t file, NSString *sourceFilePath, NSString *destinationFilePath)
 {
     NSString *patchFile = temporaryFilename(@"apply-binary-delta");
     xar_extract_tofile(x, file, [patchFile fileSystemRepresentation]);
     const char *argv[] = {"/usr/bin/bspatch", [sourceFilePath fileSystemRepresentation], [destinationFilePath fileSystemRepresentation], [patchFile fileSystemRepresentation]};
-    bspatch(4, (char **)argv);
+    BOOL success = (bspatch(4, (char **)argv) == 0);
     unlink([patchFile fileSystemRepresentation]);
+    return success;
 }
 
 int applyBinaryDelta(NSString *source, NSString *destination, NSString *patchFile)
 {
-    xar_t x = xar_open([patchFile UTF8String], READ);
+    xar_t x = xar_open([patchFile fileSystemRepresentation], READ);
     if (!x) {
-        fprintf(stderr, "Unable to open %s. Giving up.\n", [patchFile UTF8String]);
+        fprintf(stderr, "Unable to open %s. Giving up.\n", [patchFile fileSystemRepresentation]);
         return 1;
     }
+    
+    SUBinaryDeltaMajorVersion majorDiffVersion = FIRST_DELTA_DIFF_MAJOR_VERSION;
+    SUBinaryDeltaMinorVersion minorDiffVersion = FIRST_DELTA_DIFF_MINOR_VERSION;
```

Comment:
Having a way to tell clients that are too old to give up is useful, but it's a binary decision: the patch will either work well enough to be acceptable, or won't.

> For example, before this PR I made a couple backwards compatible changes for creating the diffs: case sensitive file renaming issue and a failing case where a directory became a symbolic link. For these cases, the minor version would be bumped (and in this patch I bumped it to 1.1 when creating version 1 patches).

But it works fine without clients knowing about this. 1.0 client can't do anything with that information, so the patch applied by 1.0 must be correct by definition. 

1.1 client has a choice of using its 1.1 behavior (for which it doesn't need minor version information) or emulating 1.0, so that patch is applied in a bug-compatible way.

I find bug-emulation unappealing and try to avoid it, but if was necessary, it could be achieved by adding an optional key that's a flag `"do-this-weird-buggy-thing-that-1.0-did"` that newer versions emit and recognize.

>  If I sent a version 1 patch to someone to apply, they can tell if the patch was created by version 1.0 or 1.1, which may be the difference for the patch applying correctly.

How so? If there's possibility of patch being applied incorrectly, then it should be a major version change.

In case diff format v1.1 added a new optional feature, then v1.1 patch program could look for the feature's data, rather than at the version number.

e.g. if you add `xattr` support under a new key, then v1.0 patch will apply content changes, but won't change xattrs, and v1.1 patch will apply content and xattr changes, just by looking for the xattr key - it won't need to look at the version.